File: src/main/java/com/github/nsc/de/shake/parser/node/Node.java
package com.github.nsc.de.shake.parser.node;

import com.github.nsc.de.shake.util.characterinput.position.PositionMap;

public abstract class Node {

    private final PositionMap map;

    protected Node(PositionMap map) {
        this.map = map;
    }

    public PositionMap getMap() {
        return map;
    }
}

File: parser/src/main/kotlin/com/github/nsc/de/shake/parser/node/Node.kt
package com.github.nsc.de.shake.parser.node

import com.github.nsc.de.shake.util.characterinput.position.PositionMap
import com.github.nsc.de.shake.util.json.JSON

abstract class Node protected constructor(val map: PositionMap) {

    val json: Map<String, *>
        get() = toJson()

    abstract fun toJson(): Map<String, *>
    override fun toString(): String = JSON.stringify(this.json)

}